### ISSUE TITLE
The meat for cooking is taken from a working smoking rack

### ISSUE THREAD

**Issue by IdleSol** [CleverRaven/Cataclysm-DDA#73173]
### Describe the bug

<s>Fried meat. I have 4 chunks of meat that I can roast an infinite number of times. Probably related to this topic #69507 </s>

I assume the mistake will be repeated not only with meat.

### Attach save file

[Госпорт-trimmed.tar.gz](https://github.com/CleverRaven/Cataclysm-DDA/files/15052145/-trimmed.tar.gz)

### Steps to reproduce

1. Load the save.
2. Open recent recipes
3. Repeat the fried meat recipe (selecting all 4 pieces)
4. Check the floor, there are pieces of meat still on the floor.
![1](https://github.com/CleverRaven/Cataclysm-DDA/assets/155764584/b0d95199-4cf8-4d2b-8bb0-0d60d095ea5c)

<s>On a clean game, the key is the meat lying on the floor. Haven't checked it, judging from past threads.</s>

UPD. Meat is wasted from a working smoking rack.

5. Go through the door to the street. There are three smokehouses located there. The middle one is working
![2](https://github.com/CleverRaven/Cataclysm-DDA/assets/155764584/f3a280b9-c8ff-4563-9dc0-f52f4dcd643e)
6. Put out the smoking rack and take out the meat
7. Memorize the amount of meat
8. Start the smoker
9. Start cooking the roast meat (you do not have to finish it)
10. Repeat steps 6 and 7.
11. Compare the amount of meat

### Expected behavior

Cooking should consume the components.

### Versions and configuration

- OS: Windows
    - OS Version: 10.0.19043 (21H1)
- Game Version: cdda-experimental-2024-04-21-0738 0a3e18c [64-bit]
- Graphics Version: Tiles
- Game Language: Русский <color_dark_gray>(90.0%)</color> [ru]
- Mods loaded: [
    Тёмные Дни Впереди [dda],
    Отключить потребности у НПС [no_npc_food],
    Портальные штормы игнорируют НПС [personal_portal_storms],
    Медленный рост фунгалоидов [no_fungal_growth]
]

### Additional context

Meat scraps, spent as they should be. Lungs, liver, too. Chunks of fat, no mistakes either.

**Comment by NetSysFire** [comment]
Both issues really do sound like a regression. Do you happen to know when approximately this bug was introduced? That could speed up the fix.

**Comment by IdleSol** [comment]
I tried reproducing on a new save. And it didn't work. I started looking to see what the difference was. I found out that they are different errors.

In this case, the meat is wasted. But it's taken from a working smoking rack.

**Comment by IdleSol** [comment]
Yeah, I tested it on corn kernels.

**Comment by osuphobia** [comment]
A sixth sense told me that this bug must be introduced when removing charges from food stuff.
Tested in https://github.com/CleverRaven/Cataclysm-DDA/releases/tag/cdda-experimental-2023-08-02-0307 and reproduced it, while things worked as intended in https://github.com/CleverRaven/Cataclysm-DDA/releases/tag/cdda-experimental-2023-07-31-0609.
Then I tried https://github.com/CleverRaven/Cataclysm-DDA/releases/tag/cdda-experimental-2023-08-02-0110, and the meat chunks out of the rack were consumed, so the reason of this issue should be https://github.com/CleverRaven/Cataclysm-DDA/pull/67356.

**Comment by NetSysFire** [comment]
CC @mqrause then, the author of the linked PR. Maybe they know what is happening here.

**Comment by mqrause** [comment]
> A sixth sense told me that this bug must be introduced when removing charges from food stuff. Tested in https://github.com/CleverRaven/Cataclysm-DDA/releases/tag/cdda-experimental-2023-08-02-0307 and reproduced it, while things worked as intended in https://github.com/CleverRaven/Cataclysm-DDA/releases/tag/cdda-experimental-2023-07-31-0609. Then I tried https://github.com/CleverRaven/Cataclysm-DDA/releases/tag/cdda-experimental-2023-08-02-0110, and the meat chunks out of the rack were consumed, so the reason of this issue should be #67356.

I'm not sure I understand the conclusion. If experimental 2023-08-02-0307 merged that PR, but 2023-08-02-0110 already had the issue, the cause should be something different.

**Comment by osuphobia** [comment]
>the meat chunks out of the rack were consumed

Sorry, maybe this sentence is misleading. I mean in 2023-08-02-0110 this issue is not existing, the chunks of meat in the smoking rack did not disapper, and the chunks outside were consumed correctly when crafting cooked meat.

**Comment by osuphobia** [comment]
https://github.com/CleverRaven/Cataclysm-DDA/assets/78858975/915d5fb8-151b-41b9-9398-3fb368e66b15

In 2023-08-02-0110 one chunk outside is consumed (total amount is 101 as the rack duplicated one chunk).


https://github.com/CleverRaven/Cataclysm-DDA/assets/78858975/b676da11-a6b6-4c2a-bf08-f40a3bdde628

In 2023-08-02-0307 the chunk in the rack is consumed (the duplicate bug is fixed so total amount is 100).

**Comment by mqrause** [comment]
That's curious. The PR in question just changes the charges count of the items put into the rack. That shouldn't have any effect on component selection for crafting. I suspect the real bug is hiding somewhere in the construction of the crafting inventory and that just surfaced it somehow.

**Comment by osuphobia** [comment]
Maybe you are right, I'll try to figure it out.
Just found something quite interesting.

If you migrate a save before 2023-08-02-0307 that already has a smoking rack, chunks in it will not be wrongly consumed.
But a new smoking rack built after migrating to 2023-08-02-0307 and newer version causes this issue again.
[Wakarusa-trimmed.tar.gz](https://github.com/CleverRaven/Cataclysm-DDA/files/15059331/Wakarusa-trimmed.tar.gz)

**Comment by IdleSol** [comment]
You don't cook meat right. More specifically, you have the wrong kitchen. You need one like this:
![test](https://github.com/CleverRaven/Cataclysm-DDA/assets/155764584/2c69e2a0-e8e8-47f0-84ed-1b14fa41cce8)

Index 1. brazier + plank
Index 2. smoking rack (12к charcoal + 80 chunk of meat) (*Note. I assume charcoal is optional, but I tested with it.*)
Index 3. brazier + plank (*Note. Optional. I used for symmetry to eliminate some options.*)
Index 4..6 One chunk of meat

You need the following versions of the game:
- [2023-07-31-0609](https://github.com/CleverRaven/Cataclysm-DDA/releases/tag/cdda-experimental-2023-07-31-0609)
- [2023-07-31-1032](https://github.com/CleverRaven/Cataclysm-DDA/releases/tag/cdda-experimental-2023-07-31-1032)

I don't know how they react to saves. In each version, I started with a clean character and options.

**Steps to reproduce**

1. Build a fire in the braziers
2. Create a save
3. Start cooking the meat according to the recipe for `cooked meat`. One chunk at a time. (You need a pot or a frying pan.)
4. Look at the meat lying on the ground
а. Versions 2023-07-31-0609. Meat is spent sequentially: 4-5-6. After the meat from point 5 has been spent. The recipe becomes inactive.
b. Versions 2023-07-31-1032. Meat is spent sequentially: 4-5. But at point 6, no meat is wasted. It is taken from the smoking rack

You can load from a save and try different character positions. At the top brazier, at the bottom brazier. Or anywhere else. The meat at point 6 is not wasted.

**To summarize**

Formally, the bug appeared after switching from charges to individual pieces. #60885 
@osuphobia You guessed the "author" @mqrause But I highly doubt that the problem is in fixing it.

Rather in the algorithm of searching for components for crafting. He looks for them from right to left, starting from the upper right corner. And when he reaches the rack, he finds them.

A possible fix would be to mark the contents of the rack as ignored components.

The only question is what else might qualify. Water in a washing machine? In an autoclave? Charcoal in a smoking rack when creating steel? That is, does it only apply to the smoking rack or is it a more general problem.

**Comment by IdleSol** [comment]
Tested the water in a home washing machine. It is not wasted in making clean water.

**Comment by osuphobia** [comment]
@IdleSol 
Tested again, you are partly right.
Tools and fuels are irrelevant (for cooked meat shovel is enough), the game just attempts to consume components from the top left to bottom right, searching row by row. (You said right to left)
![z1x2](https://github.com/CleverRaven/Cataclysm-DDA/assets/78858975/de06ca10-476a-4ed9-8732-87351e567d89)
Functions for crafting seems to treat items with/without charge differently, I'm still viewing them.

**Comment by osuphobia** [comment]
>The only question is what else might qualify. Water in a washing machine? In an autoclave? Charcoal in a smoking rack when creating steel? That is, does it only apply to the smoking rack or is it a more general problem.

Tested wood keg and wood fermenting vat (added brewable data to water so it can be poured into the fermenting vat).
Water in the keg will be consumed when crafting clean water, while water in the vat will not. Water in the keg is counted when checking the recipe in the crafting gui.

So we got three conditions:

1. Smoking rack. Components in it will not be counted in the crafting gui, but will be consumed in crafting command.
2. Wood keg. Components in it will **both** be counted in the crafting gui and consumed in crafting command.
3. Fermenting vat.  Components in it will **nether** be counted in the crafting gui nor be consumed in crafting command.

For charcoal in a smoking rack, it will not be consumed, as it is loaded in the fake_char_smoker pseudo tool.
Washing machine and autoclave are Appliance/Power Grid things, did not check that part of functions.
